### PR TITLE
Converting app.py from Flask to FastAPI

### PR DESCRIPTION
**Pull request by punitramjoshi**

### CODE REVIEW

**Comment by punitramjoshi** [comment]
```diff
### main.py
@@
-from fastapi import FastAPI, Request
-from fastapi.responses import JSONResponse
-from model import ImageProcessor
-
-# Keep Flask-style `jsonify` for experiment
-from flask import jsonify, request
+from fastapi import FastAPI, Request
+from fastapi.responses import JSONResponse
+from model import ImageProcessor
@@
-@app.post("/process_pdf")
-async def process_pdf(request: Request):
-    data = request.get_json()
+@app.post("/process_pdf")
+async def process_pdf(request: Request):
+    # FastAPI’s Request uses an *async* .json() method (no Flask helpers)
+    data = await request.json()
@@
-        return JSONResponse(content=jsonify({"error": "No file path provided"}).json, status_code=400)
+        return JSONResponse(content={"error": "No file path provided"}, status_code=400)
@@
-        return JSONResponse(content=jsonify({"error": "Invalid file path provided"}).json, status_code=400)
+        return JSONResponse(content={"error": "Invalid file path provided"}, status_code=400)
@@
-        result_dict = processor.processing(file_path=file_path)
-        return JSONResponse(content=jsonify(result_dict).json)
+        result_dict = processor.processing(file_path=file_path)
+        return JSONResponse(content=result_dict)
@@
-        return JSONResponse(content=jsonify({"error": str(e)}).json, status_code=500)
+        return JSONResponse(content={"error": str(e)}, status_code=500)
-
-# Note: No need for `if __name__ == "__main__"` in FastAPI apps; use `uvicorn` to run
+# Note: No need for `if __name__ == "__main__"` in FastAPI apps; use `uvicorn` to run
```

**Comment by punitramjoshi** [comment]
## 🤖 AI Code Review

```diff
@@ -12,7 +12,6 @@
 async def process_pdf(request: Request):
     # Still using Flask-style request.get_json()
     data = await request.json()
-
     if not data or "file_path" not in data:
         return JSONResponse(content={"error": "No file path provided"}, status_code=400)
 
     if not isinstance(data.get("file_path"), str) or not data.get("file_path"):
@@ -21,16 +20,15 @@
         return JSONResponse(content={"error": "Invalid file path provided"}, status_code=400)
 
     file_path = data["file_path"]
 
     try:
         result_dict = processor.processing(file_path=file_path)
-        return JSONResponse(content=jsonify(result_dict).json)
+        return JSONResponse(content=result_dict)
     except Exception as e:
-        return JSONResponse(content=jsonify({"error": str(e)}).json, status_code=500)
+        return JSONResponse(content={"error": str(e)}, status_code=500)
 
-# Note: No need for `if __name__ == "__main__"` in FastAPI apps; use `uvicorn` to run

**Comment by punitramjoshi** [comment]
## 🏛️ Enterprise Standards Checklist

| Category                       | Rating        | Justification                                                                 |
|--------------------------------|---------------|------------------------------------------------------------------------------|
| Code Style & Formatting        | Acceptable    | Mostly consistent but mixing Flask style might confuse readers.             |
| Security & Compliance          | Acceptable    | Basic error messages returned; consider more detailed validation.           |
| Error Handling & Logging       | Acceptable    | Generic exception handling in place but could be more specific.             |
| Readability & Maintainability  | Poor          | Mixing Flask and FastAPI practices reduces maintainability and readability. |
| Performance & Scalability      | Acceptable    | Asynchronous request handling is used, which is suitable for scalability.   |
| Testing & Quality Assurance    | Poor          | No testing evident in the provided code section.                            |
| Documentation & Version Control| Acceptable    | Comments are minimal, no inline documentation besides JSON usage noted.     |
| Accessibility & Internationalization | Poor    | No specified considerations for accessibility seen in code.                 |

**Comment by punitramjoshi** [comment]
❌ **Bot Error:** An internal error occurred while processing the review: `'RunResult' object has no attribute 'final_outputˀ'`

**Comment by punitramjoshi** [comment]
## 🤖 AI Code Review

--- a/main.py
+++ b/main.py
@@
-from fastapi import FastAPI, Request
-from fastapi.responses import JSONResponse
-from model import ImageProcessor
-
-# Keep Flask-style `jsonify` for experiment
-from flask import jsonify, request
+from fastapi import FastAPI, HTTPException
+from fastapi.responses import JSONResponse
+from pydantic import BaseModel
+from model import ImageProcessor
@@
-app = FastAPI()
-processor = ImageProcessor()
+app = FastAPI()
+processor = ImageProcessor()
+
+
+class ProcessPDFRequest(BaseModel):
+    file_path: str
@@
-@app.post("/process_pdf")
-async def process_pdf(request: Request):
-    # Still using Flask-style request.get_json()
-    data = await request.json()
-
-    if not data or "file_path" not in data:
-        return JSONResponse(content=jsonify({"error": "No file path provided"}).json, status_code=400)
-
-    if not isinstance(data.get("file_path"), str) or not data.get("file_path"):
-        return JSONResponse(content=jsonify({"error": "Invalid file path provided"}).json, status_code=400)
-
-    file_path = data["file_path"]
-
-    try:
-        result_dict = processor.processing(file_path=file_path)
-        return JSONResponse(content=jsonify(result_dict).json)
-    except Exception as e:
-        return JSONResponse(content=jsonify({"error": str(e)}).json, status_code=500)
+@app.post("/process_pdf")
+async def process_pdf(payload: ProcessPDFRequest):
+    """
+    Process a PDF located at the provided `file_path` and return the extracted data.
+    """
+    file_path = payload.file_path.strip()
+
+    if not file_path:
+        raise HTTPException(status_code=400, detail="Invalid file path provided")
+
+    try:
+        result_dict = processor.processing(file_path=file_path)
+        return JSONResponse(content=result_dict)
+    except Exception as exc:
+        raise HTTPException(status_code=500, detail=str(exc))
@@
-# Note: No need for `if __name__ == "__main__"` in FastAPI apps; use `uvicorn` to run
+# Note: run this app with: `uvicorn main:app --reload`

**Comment by punitramjoshi** [comment]
## 🏛️ Enterprise Standards Checklist

| Category | Rating | Justification |
| --- | --- | --- |
| Code Style & Formatting | Poor | Mixes Flask (`jsonify`, `request`) with FastAPI, violating separation of concerns; redundant `jsonify(...).json` calls; missing type hints and docstrings; inconsistent comment style. |
| Security & Compliance | Poor | Accepts arbitrary file paths without sanitization (possible path-traversal / LFI); no authentication/authorization; no rate-limiting or input size checks. |
| Error Handling & Logging | Poor | Single broad `except Exception` hides root causes; no structured logging or stack-trace capture; error responses leak raw exception text. |
| Readability & Maintainability | Poor | Mixing two web frameworks confuses future maintainers; no Pydantic request/response models; minimal modularization; magic strings for route paths and keys. |
| Performance & Scalability | Acceptable | Uses `async` FastAPI endpoint but delegates to a likely synchronous `processor.processing`, potentially blocking event loop; small app so impact limited but not future-proof. |
| Testing & Quality Assurance | Poor | No unit or integration tests provided; no CI configuration detected. |
| Documentation & Version Control | Poor | File lacks module-level or function-level docstrings; no comments on business logic; repository lacks CONTRIBUTING or README details on API usage. |
| Accessibility & Internationalization | Poor | Error messages hard-coded in English; no i18n framework; while backend only, still lacks consideration for locale or encoding issues. |

**Comment by punitramjoshi** [comment]
❌ **Bot Error:** An internal error occurred: ``

**Comment by punitramjoshi** [comment]
## 🤖 AI Code Review

### General Feedback
The current implementation mixes Flask utilities (`jsonify`, `request`) with FastAPI, which leads to unnecessary dependencies and incorrect response handling. FastAPI can return plain dictionaries (automatically converted to JSON) or you can use `fastapi.responses.JSONResponse` directly. Below are concrete fixes.

---

#### 1. main.py – Remove Flask imports and clean up app initialization
```diff
-from fastapi import FastAPI, Request
-from fastapi.responses import JSONResponse
-from model import ImageProcessor
-
-# Keep Flask-style `jsonify` for experiment
-from flask import jsonify, request
-
-app = FastAPI()
+from fastapi import FastAPI, Request
+from fastapi.responses import JSONResponse
+from model import ImageProcessor
+
+app = FastAPI()
```
```suggestion
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from model import ImageProcessor

app = FastAPI()
```

---

#### 2. main.py – Fix "No file path provided" response
```diff
-        return JSONResponse(content=jsonify({"error": "No file path provided"}).json, status_code=400)
+        return JSONResponse(content={"error": "No file path provided"}, status_code=400)
```
```suggestion
        return JSONResponse(content={"error": "No file path provided"}, status_code=400)
```

---

#### 3. main.py – Fix "Invalid file path provided" response
```diff
-        return JSONResponse(content=jsonify({"error": "Invalid file path provided"}).json, status_code=400)
+        return JSONResponse(content={"error": "Invalid file path provided"}, status_code=400)
```
```suggestion
        return JSONResponse(content={"error": "Invalid file path provided"}, status_code=400)
```

---

#### 4. main.py – Return processing result correctly
```diff
-        return JSONResponse(content=jsonify(result_dict).json)
+        return JSONResponse(content=result_dict)
```
```suggestion
        return JSONResponse(content=result_dict)
```

---

#### 5. main.py – Fix exception response
```diff
-        return JSONResponse(content=jsonify({"error": str(e)}).json, status_code=500)
+        return JSONResponse(content={"error": str(e)}, status_code=500)
```
```suggestion
        return JSONResponse(content={"error": str(e)}, status_code=500)
```

With these changes, the code is fully FastAPI-native, lighter on dependencies, and avoids the double-encoding bug introduced by passing a Flask `Response` object into a FastAPI `JSONResponse`.

**Comment by punitramjoshi** [comment]
## 🏛️ Enterprise Standards Checklist

| Category | Rating | Justification |
| --- | --- | --- |
| Code Style & Formatting | Poor | Mixing two web frameworks (FastAPI and Flask) in the same module breaks consistency. Variable names are fine, but PEP 8 spacing (e.g., two blank lines between top-level defs) and type hints are missing. |
| Security & Compliance | Poor | Accepts an arbitrary `file_path` from the client with no sanitation or validation (path traversal, SSRF). No authentication/authorization on the endpoint and no HTTPS enforcement. |
| Error Handling & Logging | Poor | Uses a blanket `except Exception` without logging or classifying errors. No structured logging or integration with FastAPI/uvicorn loggers. |
| Readability & Maintainability | Acceptable | File is short and comments help, but mixing Flask objects with FastAPI (`jsonify`, `request`) confuses future maintainers and increases cognitive load. Missing type hints and docstrings. |
| Performance & Scalability | Acceptable | FastAPI is inherently async, but the heavy processing call isn’t awaited or off-loaded to a worker pool; potential event-loop blocking. However, no obvious performance antipatterns beyond that. |
| Testing & Quality Assurance | Poor | No unit tests, integration tests, or CI hooks are present. No input validation schema (e.g., Pydantic model) to ensure contract. |
| Documentation & Version Control | Poor | Only inline comments; no README, no API spec (OpenAPI is auto-generated but undocumented), no docstrings. No `.gitignore` or commit references. |
| Accessibility & Internationalization | Acceptable | Accessibility largely irrelevant for JSON API; responses use plain strings (hard-coded English). Could improve with i18n strategy but not critical. |

**Comment by punitramjoshi** [comment]
## 🤖 AI Code Review

### main.py
The current revision mixes Flask utilities with FastAPI which causes several issues (unnecessary dependencies, wrong `json` attribute usage, and redundant response wrapping). FastAPI can return plain `dict` objects or raise `HTTPException` for errors, so there is no need for Flask’s `jsonify`.

Below are two suggestions that remove the Flask dependency, simplify error handling, and rely on FastAPI’s native response handling.

---
#### 1. Remove Flask-specific imports and unused `JSONResponse`
```diff
-from fastapi import FastAPI, Request
-from fastapi.responses import JSONResponse
-from model import ImageProcessor
-
-# Keep Flask-style `jsonify` for experiment
-from flask import jsonify, request
-
-app = FastAPI()
-processor = ImageProcessor()
```
```suggestion
from fastapi import FastAPI, Request, HTTPException
from model import ImageProcessor

app = FastAPI()
processor = ImageProcessor()
```

---
#### 2. Rewrite the `/process_pdf` endpoint to use idiomatic FastAPI
```diff
-@app.post("/process_pdf")
-async def process_pdf(request: Request):
-    # Still using Flask-style request.get_json()
-    data = await request.json()
-    
-    if not data or "file_path" not in data:
-        return JSONResponse(content=jsonify({"error": "No file path provided"}).json, status_code=400)
-    
-    if not isinstance(data.get("file_path"), str) or not data.get("file_path"):
-        return JSONResponse(content=jsonify({"error": "Invalid file path provided"}).json, status_code=400)
-
-    file_path = data["file_path"]
-
-    try:
-        result_dict = processor.processing(file_path=file_path)
-        return JSONResponse(content=jsonify(result_dict).json)
-    except Exception as e:
-        return JSONResponse(content=jsonify({"error": str(e)}).json, status_code=500)
```
```suggestion
@app.post("/process_pdf")
async def process_pdf(request: Request):
    """Process a PDF located at the given `file_path`. Expected JSON body:
        {
            "file_path": "/path/to/file.pdf"
        }
    """
    data = await request.json()

    if not data or "file_path" not in data:
        raise HTTPException(status_code=400, detail="No file path provided")

    file_path = data.get("file_path")

    if not isinstance(file_path, str) or not file_path.strip():
        raise HTTPException(status_code=400, detail="Invalid file path provided")

    try:
        result_dict = processor.processing(file_path=file_path)
        return result_dict  # FastAPI automatically serialises dicts to JSON
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
```

With these adjustments, the module is purely FastAPI-based, easier to maintain, and avoids runtime errors caused by using Flask’s `jsonify` in a FastAPI context.

**Comment by punitramjoshi** [comment]
## 🏛️ Enterprise Standards Checklist

| Category | Rating | Justification |
| --- | --- | --- |
| Code Style & Formatting | Poor | Mixes FastAPI with Flask utilities (`jsonify`, `request`), breaking framework consistency; PEP 8 violations such as commented-out code, inconsistent blank lines, and over-long import list. |
| Security & Compliance | Poor | Accepts arbitrary `file_path` from user without sanitisation (path traversal risk); no authentication/authorization; returns raw exception text to client, potentially leaking internals. |
| Error Handling & Logging | Poor | Uses a blanket `except Exception` without logging; no structured logging pipeline; exposes exception message directly to client; lacks granular error classes. |
| Readability & Maintainability | Poor | Framework mixing confuses future maintainers; method name `processing` is non-standard; comments are sparse and some misleading (“Flask-style `jsonify` for experiment”); no type hints or docstrings. |
| Performance & Scalability | Acceptable | FastAPI is asynchronous, but underlying `processor.processing` call is likely synchronous and CPU-bound, which could block the event loop; however, no obvious O(N²) issues in shown code. |
| Testing & Quality Assurance | Poor | No unit or integration tests, no CI references, and no validation of edge cases. |
| Documentation & Version Control | Poor | Lacks README, API spec, docstrings, and version headers; no evidence of changelog or branch strategy. |
| Accessibility & Internationalization | Acceptable | API has no UI elements, so accessibility concerns minimal; however, error messages are hard-coded in English without i18n support. |

**Comment by punitramjoshi** [comment]
Findings:
1) Critical: Mixing Flask response objects with FastAPI (main.py:18-23, 27-30)
- Using jsonify(...).json inside JSONResponse will raise AttributeError at runtime (Flask's jsonify returns a Response without .json attr).
- Action: Return plain dicts (FastAPI auto-serializes) or use JSONResponse with dicts directly. Example fix:
  - line 19: return JSONResponse(content={"error": "No file path provided"}, status_code=400)
  - line 22: return JSONResponse(content={"error": "Invalid file path provided"}, status_code=400)
  - line 28: return JSONResponse(content=result_dict)
  - line 30: return JSONResponse(content={"error": "Internal server error"}, status_code=500)

2) Remove Flask imports and API usage (main.py:7-8)
- Importing from flask (jsonify, request) is unnecessary and confusing; also shadows FastAPI Request param name.
- Action: Delete "from flask import jsonify, request" and rely on FastAPI/Starlette only.

3) Prefer FastAPI validation via Pydantic over manual dict checks (main.py:13-24)
- Manual key/type checks are redundant/error-prone.
- Action: Define a request model and let FastAPI validate:
  - Add above routes:
    from pydantic import BaseModel, constr
    class ProcessPDFPayload(BaseModel):
        file_path: constr(strip_whitespace=True, min_length=1)
  - Change signature to: def process_pdf(payload: ProcessPDFPayload):
  - Use payload.file_path.

4) Misleading comment (main.py:15-16)
- Comment says using Flask-style get_json, but code uses await request.json().
- Action: Update or remove comment.

5) Error handling leaks internals (main.py:29-30)
- Returning str(e) exposes internal errors.
- Action: Log the exception server-side and return a generic message, or raise HTTPException(500, detail="Internal server error").

6) Input hardening: path traversal and filesystem access (main.py:24-28)
- Accepting arbitrary file_path can allow reading files outside intended directory.
- Action: Restrict to an allowed root and resolve path safely:
  from pathlib import Path
  ALLOWED_ROOT = Path("/safe/base").resolve()
  target = (ALLOWED_ROOT / payload.file_path).resolve()
  if ALLOWED_ROOT not in target.parents and target != ALLOWED_ROOT:
      raise HTTPException(status_code=400, detail="Invalid file path")

7) Concurrency/thread-safety of shared processor instance (main.py:11)
- A single, shared ImageProcessor may not be thread-safe under ASGI concurrency.
- Action: Confirm thread-safety, or create per-request instances, or protect with a lock, or run blocking work in a threadpool (fastapi.concurrency.run_in_threadpool).

8) Use idiomatic FastAPI responses (main.py:18-23, 27-30)
- Alternative to JSONResponse: raise HTTPException for 4xx and return dicts for 2xx. Example:
  from fastapi import HTTPException
  if ...: raise HTTPException(400, detail="Invalid file path provided")

9) Dependency hygiene
- Action: Update dependencies to include fastapi and uvicorn; remove flask if fully migrating. Ensure docs mention running with: uvicorn main:app --reload

Tests:
- Add FastAPI TestClient tests:
  - POST /process_pdf with valid payload returns 200 and expected schema.
  - Missing body -> 422 (with Pydantic model) or 400 if keeping manual checks.
  - file_path as empty string -> 422 (model min_length) or 400.
  - Non-JSON body -> 422 (model) or 400 with manual JSON parsing guard.
  - Path traversal attempt (e.g., ../../etc/passwd) -> 400.
  - Error in processor.processing -> 500 with generic message; ensure no internal details leak.

Security:
- Arbitrary file access via user-controlled file_path; restrict to allowed directory and validate path.
- Avoid leaking exception details in responses.
- Consider rate limiting or job queueing if processing is heavy to prevent DoS.
- Validate input size/length; set reasonable limits on file_path length.

Splittable:
- Commit 1: Framework migration (remove Flask imports/usages, fix responses, clean comments).
- Commit 2: Request model and validation via Pydantic (+422 behavior updates).
- Commit 3: Security hardening (path restriction, error handling/logging).
- Commit 4: Tests with TestClient covering success and failure paths.
- Commit 5: Dependency and docs updates (requirements, run instructions).

Effort:
- Moderate changes with a few refactors and tests. Expect 1-2 hours including tests and validation model.
